File: sdk/typescript/src/signing.ts
import { createHmac, createHash, timingSafeEqual } from 'node:crypto';

export type JsonValue = null | boolean | number | string | JsonValue[] | { [key: string]: JsonValue | undefined };

export function canonicalJson(value: unknown): string { return JSON.stringify(normalize(value)); }
function normalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalize);
  if (value && typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      const v = (value as Record<string, unknown>)[key];
      if (v !== undefined) out[key] = normalize(v);
    }
    return out;
  }
  return value;
}
export function sha256Hex(value: string | Buffer): string { return createHash('sha256').update(value).digest('hex'); }
export function hashResultPayload(payload: unknown): string { return sha256Hex(canonicalJson(payload)); }

export interface ResultSignatureInput {
  executionResultId: string;
  authorizationId: string;
  externalSystemId: string;
  status: string;
  executedAction: string;
  target: string | Record<string, unknown>;
  amountMinor?: number;
  amountCurrency?: string;
  externalReference?: string;
  paymentOrSettlementReference?: string;
  resultPayloadHash: string;
  executedAt: string;
  submittedAt: string;
}

export function resultSigningString(input: ResultSignatureInput): string {
  const target = typeof input.target === 'string' ? input.target : canonicalJson(input.target);
  return [
    input.executionResultId,
    input.authorizationId,
    input.externalSystemId,
    input.status,
    input.executedAction,
    target,
    input.amountMinor ?? '',
    input.amountCurrency ?? '',
    input.externalReference ?? '',
    input.paymentOrSettlementReference ?? '',
    input.resultPayloadHash,
    input.executedAt,
    input.submittedAt,
  ].map(String).join('\n');
}
export function signResult(input: ResultSignatureInput, secret: string): string { return createHmac('sha256', secret).update(resultSigningString(input)).digest('hex'); }
export function verifyResultSignature(input: ResultSignatureInput, signature: string, secret: string): boolean {
  let actual: Buffer;
  try { actual = Buffer.from(signature, 'hex'); } catch { return false; }
  const expected = Buffer.from(signResult(input, secret), 'hex');
  return expected.length === actual.length && timingSafeEqual(expected, actual);
}
export function attachResultSignature<T extends Omit<ResultSignatureInput, 'resultPayloadHash'> & { resultPayload?: unknown; signerKeyId: string }>(result: T, secret: string): T & { resultPayloadHash: string; signature: string } {
  const resultPayloadHash = hashResultPayload(result.resultPayload ?? {});
  const signature = signResult({ ...result, resultPayloadHash }, secret);
  return { ...result, resultPayloadHash, signature };
}

File: sdk/python/compliagl/signing.py
from __future__ import annotations

import hashlib
import hmac
import json
from dataclasses import replace
from typing import Any, Mapping

from .models import ExternalExecutionResult


def canonical_json(value: Any) -> str:
    return json.dumps(value if value is not None else {}, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def sha256_hex(value: str | bytes) -> str:
    data = value.encode("utf-8") if isinstance(value, str) else value
    return hashlib.sha256(data).hexdigest()


def result_payload_hash(result_payload: Mapping[str, Any] | None) -> str:
    return sha256_hex(canonical_json(result_payload or {}))


def canonical_signing_string(result: ExternalExecutionResult | Mapping[str, Any]) -> str:
    data = result.to_dict() if isinstance(result, ExternalExecutionResult) else dict(result)
    auth_id = data.get("authorization_id") or data.get("execution_authorization_id") or ""
    target = data.get("target")
    target_part = canonical_json(target) if isinstance(target, (dict, list)) else str(target or "")
    fields = [
        data.get("execution_result_id", ""),
        auth_id,
        data.get("external_system_id", ""),
        str(data.get("status", "")),
        data.get("executed_action", ""),
        target_part,
        "" if data.get("amount_minor") is None else str(data.get("amount_minor")),
        data.get("amount_currency") or "",
        data.get("external_reference") or "",
        data.get("payment_or_settlement_reference") or "",
        data.get("result_payload_hash", ""),
        data.get("executed_at", ""),
        data.get("submitted_at", ""),
    ]
    return "\n".join(fields)


class ResultSigner:
    def __init__(self, signing_secret: str, signer_key_id: str):
        self.signing_secret = signing_secret
        self.signer_key_id = signer_key_id

    def sign(self, result: ExternalExecutionResult) -> ExternalExecutionResult:
        return sign_execution_result(result, self.signing_secret, self.signer_key_id)


def sign_execution_result(result: ExternalExecutionResult, signing_secret: str, signer_key_id: str) -> ExternalExecutionResult:
    payload_hash = result_payload_hash(result.result_payload)
    prepared = replace(result, result_payload_hash=payload_hash, signer_key_id=signer_key_id, signature="")
    signature = hmac.new(signing_secret.encode("utf-8"), canonical_signing_string(prepared).encode("utf-8"), hashlib.sha256).hexdigest()
    return replace(prepared, signature=signature)


signExecutionResult = sign_execution_result


def verify_execution_result_signature(result: ExternalExecutionResult | Mapping[str, Any], signing_secret: str) -> bool:
    data = result.to_dict() if isinstance(result, ExternalExecutionResult) else dict(result)
    expected_hash = result_payload_hash(data.get("result_payload"))
    if not hmac.compare_digest(str(data.get("result_payload_hash", "")), expected_hash):
        return False
    actual = str(data.get("signature", ""))
    data["signature"] = ""
    expected = hmac.new(signing_secret.encode("utf-8"), canonical_signing_string(data).encode("utf-8"), hashlib.sha256).hexdigest()
    return hmac.compare_digest(actual, expected)


verifyExecutionResultSignature = verify_execution_result_signature
